daml-assistant/daml-project-config/DA/Daml/Project/Config.hs

```diff
@@ -63,6 +76,95 @@ readConfig name path = do
+        replacement <- replacer match
+        let fullMatch = head match
+        pure (replacedSoFar <> textSub haystackPosition (smStart fullMatch) haystack <> replacement, smEnd fullMatch)
+  -- Run the fold, which builds a string ending on the last match
+  (replacedToLastMatch, haystackPosition) <- foldlM stepReplace initialReplaceState matchPositions
+  -- Add the final suffix (which has no matches)
+  pure $ replacedToLastMatch <> textSub haystackPosition (T.length haystack) haystack
+
+-- First group is leading character, would ideally be a non-capturing group
+-- Second group is leading backslashes. Odd length means no replacement, even means replacement
+-- Third group is the variable name
+envVarPattern :: String
+envVarPattern = "(^|[^\\\\])(\\\\*)\\${([^}]+)}"
+
+-- Finds any ${...} where the dollar isn't preceeded by (an odd number of) '\'
+-- Replaces the inner name with the value provided by the below mapping
+interpolateEnvVariables :: [(String, String)] -> T.Text -> Either String T.Text
+interpolateEnvVariables env str = 
+  replaceAllInM envVarPattern str $ \case
+    [_, firstChar, escapeSlashes, name] -> do
+      -- Divide by 2 and floor to get the number of `\` that should be in the final string
+      let prefix = smText firstChar <> T.replicate (smLength escapeSlashes `div` 2) "\\"
+      if smLength escapeSlashes `mod` 2 == 1 -- Odd number of backslashes, this expression is escaped
+        then pure $ prefix <> "${" <> smText name <> "}"
+        else do
+          let envVarName = T.unpack $ smText name
+              mEnvVar = lookup envVarName env
+          envVarValue <- maybeToEither ("Couldn't find environment variable " <> envVarName <> " in value " <> T.unpack str) mEnvVar
+          pure $ prefix <> T.pack envVarValue
+    _ -> Left "Impossible incorrect regex submatches"
+
+-- | (internal) Helper function for defining 'readXConfig' functions.
+-- Throws a ConfigError if reading or parsing fails.
+-- Also performs environment variable interpolation on all string fields in the form of ${ENV_VAR}.
+-- TODO: ensure everything that reads the daml.yaml uses this, e.g. assistant for daml.yaml, multi-package for dependencies, etc.
+readConfigWithEnv :: Y.FromJSON b => Text -> FilePath -> IO b
+readConfigWithEnv name path = do
+  configE <- runExceptT $ do
+    (configValue :: Y.Value) <- ExceptT $ Y.decodeFileEither path
+    env <- liftIO getEnvironment
+    configValueTransformed <- 
+      except $ flip transformM configValue $ \case
+        Y.String str -> bimap Y.AesonException Y.String $ interpolateEnvVariables env str
+        v -> pure v
+    case fromJSON configValueTransformed of
+      Success x -> pure x
+      Error str -> throwE $ Y.AesonException str
+  fromRightM (throwIO . ConfigFileInvalid name) configE
+
+-- Checks if a config uses environment variable interpolation
+-- Will return False if the file does not exist, rather than throwing an error.
```

Comment:
it will also return `False` if the file _does_ exist but doesn't correctly parse as yaml. I think that's okay since the step that actually looks at the parsed output will fail, but I'd appreciate a comment to that effect :)